evm/src/arithmetic/addcc.rs

```diff
@@ -0,0 +1,384 @@
+        output[i] = s & MASK;
+    }
+    (output, cy)
+}
+
+fn u256_sub_br(input0: [u64; N_LIMBS], input1: [u64; N_LIMBS]) -> ([u64; N_LIMBS], u64) {
+    const LIMB_BOUNDARY: u64 = 1 << LIMB_BITS;
+    const MASK: u64 = LIMB_BOUNDARY - 1u64;
+
+    let mut output = [0u64; N_LIMBS];
+    let mut br = 0u64;
+    for (i, a, b) in izip!(0.., input0, input1) {
+        let d = LIMB_BOUNDARY + a - b - br;
+        // if a < b, then d < 2^16 so br = 1
+        // if a >= b, then d >= 2^16 so br = 0
+        br = 1u64 - (d >> LIMB_BITS);
+        assert!(br <= 1u64, "input limbs were larger than 16 bits");
+        output[i] = d & MASK;
+    }
+
+    (output, br)
+}
+
+/// Generate row for ADD, SUB, GT and LT operations.
+///
+/// A row consists of four values, GENERAL_REGISTER_[012] and
+/// GENERAL_REGISTER_BIT. The interpretation of these values for each
+/// operation is as follows:
+///
+/// ADD: REGISTER_0 + REGISTER_1, output in REGISTER_2, ignore REGISTER_BIT
+/// SUB: REGISTER_0 - REGISTER_2, output in REGISTER_1, ignore REGISTER_BIT
+///  GT: REGISTER_0 > REGISTER_2, output in REGISTER_BIT, auxiliary output in REGISTER_1
+///  LT: REGISTER_2 < REGISTER_0, output in REGISTER_BIT, auxiliary output in REGISTER_1
+pub(crate) fn generate<F: RichField>(lv: &mut [F], filter: usize) {
+    match filter {
+        IS_ADD => {
+            let x = read_value_u64_limbs(lv, GENERAL_REGISTER_0);
+            let y = read_value_u64_limbs(lv, GENERAL_REGISTER_1);
+
+            // x + y == z + cy*2^256
+            let (z, cy) = u256_add_cc(x, y);
+
+            lv[GENERAL_REGISTER_2].copy_from_slice(&z.map(F::from_canonical_u64));
+            lv[GENERAL_REGISTER_BIT] = F::from_canonical_u64(cy);
+        }
+        IS_SUB | IS_GT | IS_LT => {
+            let x = read_value_u64_limbs(lv, GENERAL_REGISTER_0);
+            let z = read_value_u64_limbs(lv, GENERAL_REGISTER_2);
+
+            // y == z - x + cy*2^256
+            let (y, cy) = u256_sub_br(z, x);
+
+            lv[GENERAL_REGISTER_1].copy_from_slice(&y.map(F::from_canonical_u64));
+            lv[GENERAL_REGISTER_BIT] = F::from_canonical_u64(cy);
+        }
+        _ => panic!("unexpected operation filter"),
+    };
+}
+
+/// Given a polynomial f represented by its coefficients `pol_coeffs`
+/// and a number n represented by its digits `digits` in base
+/// B=2^LIMB_BITS, with `pol_coeffs` and `digits` being of equal
+/// length as sequences, this method constrains that f(B) == n, but
+/// allowing the possibility of overflow by one bit (which is
+/// returned).
+fn eval_packed_generic_pol_eval_equal<P, I, J>(
+    yield_constr: &mut ConstraintConsumer<P>,
+    filter: P,
+    pol_coeffs: I,
+    digits: J,
+    is_two_row_op: bool,
+) -> P
+where
+    P: PackedField,
+    I: Iterator<Item = P>,
+    J: Iterator<Item = P>,
+{
+    let overflow = P::Scalar::from_canonical_u64(1 << LIMB_BITS);
+    let overflow_inv = overflow.inverse();
```

Comment:
nit: There's a `inverse_2exp` that could be used here. But since this is called quite often, maybe it makes sense to just have the inverse as a `const`?